map_gen/maps/danger_ores/modules/biter_drops_exotic_industries.lua

```diff
@@ -0,0 +1,69 @@
+local Event = require 'utils.event'
+local Task = require 'utils.task'
+local Token = require 'utils.token'
+local table = require 'utils.table'
+local Random = require 'map_gen.shared.random'
+
+local seed1 = 1410
+local seed2 = 12900
+local random = Random.new(seed1, seed2)
+
+local weights = {
+  ['ei_energy-crystal'] =  10,
+  ['ei_gold-chunk']     =  10,
+  ['ei_alien-resin']    = 150,
+  ['ei_alien-seed']     =   4,
+  ['ei_alien-beacon']   =   1,
+}
+
+local function draw_random()
+  local weighted_table = {}
+  local total = 1
+  for item, weight in pairs(weights) do
+    for _=0, weight+1 do
+      weighted_table[total] = item
+      total = total + 1
+    end
+  end
+
+  table.shuffle_table(weighted_table)
+
+  return weighted_table[random:next_int(1, total)]
+end
```

Comment:
I don't think using `map_gen.shared.random` like that is desync safe in multiplayer. I think this can be easy converted to use `math.random` instead.

Also in my suggestion I changed the for loop range to `1, weight` because for loop ranges are inclusive so `0, weight+1` would be two extra?

```suggestion
local weights = {
  ['ei_energy-crystal'] =  10,
  ['ei_gold-chunk']     =  10,
  ['ei_alien-resin']    = 150,
  ['ei_alien-seed']     =   4,
  ['ei_alien-beacon']   =   1,
}

local weighted_table = {}
for item, weight in pairs(weights) do
  for _ = 1, weight do
    weighted_table[#weighted_table + 1] = item
  end
end

local function draw_random()
  return weighted_table[math.random(#weighted_table)]
end
```

Reply:
Yeah it totally makes sense. I don't even know where the 0/+1 came from